Clean up cell selection behavior

jasongrout opened the pull request:
This PR:

* disambiguates *selected* from *active* at the notebook widget level. There is a new function `isSelectedOrActive` that is the equivalent of the previous `isSelected`, and now `isSelected` and friends only refer to the selected property of a cell.
* introduces a contiguous cell selection framework on the notebook widget. A contiguous selection is a single contiguous sequence of selected cells with the active cell at one end of the selection. The `head` of the selection is the active cell, and the `anchor` is the other end of the selection. Comprehensive tests are included for this framework.
* Converts all selection UX to use the contiguous selection framework
* Enhances the user interaction for selection so that it behaves more like text selection - shift-click and shift-drag updates a selection, clicking outside deselects all, etc.
* Fixes a bug in updating the active cell index when a cell is moved
* Other refactoring as necessary.

Fixes #3373.

Another PR should create comprehensive tests for the UX around selection, particularly with keyboard handling and mouse events.

blink1073 commented:
TDD baby!

blink1073 commented on packages/notebook/src/widget.ts:
```diff
@@ -846,6 +853,119 @@ class Notebook extends StaticNotebook {
+      if (index === this.activeCellIndex) {
+        // Already collapsed selection, nothing more to do.
+        return;
+      } else {
```
else after a return

jasongrout replied:
Thanks!

blink1073 commented on packages/notebook/src/widget.ts:
```diff
@@ -846,6 +853,119 @@ class Notebook extends StaticNotebook {
+        Private.selectedProperty.set(this.widgets[head], false);
+      }
+
+      // Toggle everything strictly between head and index except anchor.
```
Reverse order in comment for clarity?  `between index and head`?

blink1073 commented on test/src/notebook/widget.spec.ts:
```diff
@@ -786,32 +785,37 @@ describe('notebook/widget', () => {
+        expect(selected(widget)).to.eql([]);
+        widget.select(widget.activeCell);
+        expect(selected(widget)).to.eql([widget.activeCellIndex]);
+      })
```
Missing semicolon (from the lint report)

blink1073 commented:
Tested locally, fixed one bug where `shift+click` would immediately deselect on `mouseup`.  This LGTM, thanks!

ellisonbg commented:
@jasongrout, is this ready for testing/review?

On Thu, Dec 21, 2017 at 2:16 AM, Steven Silvester <notifications@github.com>
wrote:

> Tested locally, fixed one bug where shift+click would immediately
> deselect on mouseup. This LGTM, thanks!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyterlab/jupyterlab/pull/3414#issuecomment-353313443>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABr0E8yjaprnRNewZS4vNQMJk1iYnQHks5tCjANgaJpZM4RGTXw>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

jasongrout commented:
> Tested locally, fixed one bug where shift+click would immediately deselect on mouseup. This LGTM, thanks!

I think I accidentally overwrote your fix with a force-push (since this was still a WIP PR), but I think I fixed the same thing.

jasongrout commented:
@ellisonbg - there still are some questions about the interaction of selection and dragging, but I think the selection by itself is ready for comment. One problem I did notice with the current PR is that selecting a different notebook resets the active cell for the notebook you are leaving to 0.

jasongrout commented:
@ellisonbg - I think this is ready for taking a spin. I haven't written tests for commits past fa914e3. I can work on writing those tests, or I can work on something more pressing for the release.

ellisonbg commented:
Hmm, this builds but lab is pretty broken. Like the themes are not loading. @jasongrout have you rebased today? Could you? A lot has changed and I want to test it with everything else in place.

jasongrout commented:
Sure, I could try to rebase. I'm still not quite happy with how selection and dragging interplays, so I'm tackling that again.

jasongrout commented:
@ellisonbg - I'm much happier with how this works now. There still is a bug where it highlights too much when dragging a selection in the same document.

I'll rebase this tomorrow, but probably will not be able to do it in the morning.

jasongrout commented:
Note to self: some of the tests are out of date, and lots of new tests need to be written to clarify the behavior. Let's see how many we can get done before a release deadline :).

jasongrout commented:
(Of note, just like in a text editor, now shift and dragging the mouse changes the selection)

jasongrout commented:
The bug that I'm still tracking down is if I highlight a bunch of cells and drag to move them in the same document, the active cell after the move seems to be randomly set - sometimes it's at the top, sometimes at the bottom, and sometimes in the middle (!).

I think it's a race condition with the cell widgets catching up to the model moves. I'll investigate more when I'm back online in a few hours.

jasongrout commented:
I rebased on master.

blink1073 commented:
LGTM, no longer WIP?

jasongrout commented:
No longer WIP. Hopefully the tests pass (they do on my computer).

jasongrout commented:
Travis passes!

blink1073 commented:
Great work, thank you!
